New: Pagination Support for Devices and Volumes

ericrpowers opened the pull request:
In using the `packngo` API, I noticed pagination is not supported causing some devices to not show up whenever there is over 10. With the way we delete devices (by doing a substring match of hostname), this causes "leaking" of devices.

I also see a couple threads (#63 and #51) talk about this and one PR (#52) but no code directly tackling the pagination **and** `per_page` size.

Decided to take a stab at it for the `List` function within `devices.go`. This is more of a mocked up example of what _could_ be propagated through the code base where appropriate.

Tried to incorporate most of the comments/practices in #63 and #52.

Tested this internally with and without `per_page` and it works. If this at all looks good, I would be happy to extend it to other areas as requested.

t0mk commented:
Hey @ericrpowers thanks for the PR. I wanted to to the pagination but then got distracted by sth else.

This looks pretty good, I will take a deeper look today or tomorrow. I will also see if the acceptance tests run OK or of something needs to be fixed there.

I am thinking about how to test the pagination. Creating 11 devices is probably not a way to go, maybe 11 small volumes, or ssh keys?

t0mk commented:
@ericrpowers Maybe one thing right now:
Do you think you could modify the code so that the List method takes a *ListOptions param (a pointer to a struct), and if it's nil, it gets all of the resources.

ListOptions are defined in packngo.go.

ericrpowers commented:
Yeah, I was thinking of writing a test case for this and other extensions if we go that far, but I noticed all the test cases actually use the infra. XD

Mocking would be ideal, but beyond that we could possibly extend the logic over to a lighter resource that does not cost anything. Your suggestions are probably the ones that would be ideal. SSH keys seem the lightest of the two.

As for your comment, let me do that : )

ericrpowers commented:
Went a little overboard in trying to create a test case for testing this logic for `SSHKeyService`. Let me know if this is a bit too much for the scope.

Also, in adding the `*ListOptions` param, it made me realize this is now a breaking change. Not sure if that was expected or not, @t0mk . ; )

t0mk commented:
@ericrpowers the test looks really good actually, well done. Could you please add a test for fetching maybe page number 2, with page_size 4, and then verify that there's indeed 4 resources in the API response? you can put it in the same test function.

yes, you are right that the *ListOptions arg for the List(id string) functions break the API. However, the only way I see to do it, and not to break the API (i.e. List() methods) would be to do a new function ListPaginated(opts *ListOptions). Creating new function for the paginated fetches seems unnecessary to me, considering that to accommodate this API change means just `Devices.List(pid) => Devices.List(pid, nil)`.

I added a simple List() call to the basic device test, just so that we know that it goes through.

I am OK with merging this with pagination support for devices and SSH keys. I (or someone else) can do other resources in different PR, so this doesn't grow too big.

ericrpowers commented:
Sure thing, will add it.

Yeah, I figured it was unnecessary to have a new function. One solution to avoid it creeping into people's code unexpectedly that we use is bumping the semver to a new major release for any breaking changes. Gives a heads up that something will probably need to be handled when moving over to the new version.

Sounds good. Let me do the changes requested to wrap this up.

ericrpowers commented:
Found an old issue and a new one while trying to develop the test case for Volumes.

#### 403 (Forbidden) Parallel Resource Creation

It seems the API is not happy if one tries to make things in parallel too quickly. For example, if I try to create 11 volumes all at the same time, the create will go through successfully:
```
2018/03/28 12:44:41 Created bdaf074d-5e8d-4381-98e1-58459683dc72
2018/03/28 12:44:41 Created 5173bac0-d048-4cc7-aa8b-c05f00943f96
2018/03/28 12:44:41 Created 2941c7a0-7e3e-45b3-9ac7-a27c81a31820
2018/03/28 12:44:41 Created a2fa59a7-3c08-4f02-8921-3f3492b2f9f1
2018/03/28 12:44:42 Created fdd540d4-7f11-44b5-9da9-281bc8833dd7
2018/03/28 12:44:42 Created e6ea7548-e7af-4a73-be94-8f966d1c69e8
2018/03/28 12:44:42 Created 2a5147f7-c7fa-4ce0-aba8-0321e695046c
2018/03/28 12:44:42 Created 94922b1a-23d5-497d-8c1a-55c614bc0b1c
2018/03/28 12:44:42 Created e6e8256b-fcae-410c-b7f4-08971d39e2e4
2018/03/28 12:44:42 Created a1c6cf44-1670-4cda-8a27-f640f616ac1b
2018/03/28 12:44:44 Created 9074907a-acc6-45bd-9dcb-5f9823705683
```
Then I try to do a `GET` on them to see when they are active (by means of `waitVolumeActive`) it hits this issue:
```
GET https://api.packet.net/storage/9074907a-acc6-45bd-9dcb-5f9823705683?include=facility,snapshot_policies,attachments.device: 403 You are not authorized to view this volume 
```
Made me really confused, but then I realized when I put a pause in the test right after create, the resources show like they are being created, but then, poof, they are gone. That error above must be the reaction that the resource is either deleted/corrupted/failed on the backend making it an invalid resource to access.

> Note: I actually have been seeing this 403 issue at least 1-3 times a day over the last ~2 weeks when creating packet servers in our infra. Now that I know how to avoid it, I will make the adjustments on my own end to work around this as well.

#### Creation of Volumes with Descriptions blocks deletion

I was using the code from `TestAccVolume` to create the `VolumeCreateRequest`, but when I caused a panic before teardown, I noticed leftover volumes. In the process of manually deleting them through the UI, I was presented with something along the lines of 'Please enter "null" to delete this volume'. I tried `null`, `"null"`, and empty with no luck. I had to add a description to the volume for me to be allowed to delete the volume through the UI. The API has a way, but it was a strange inconsistency that I needed the description of the volume to delete through the UI rather than the name of the volume which seems like a requirement.

@t0mk, if these are new issues and you would like me to report them somewhere, just let me know. : )

ericrpowers commented:
I had to remove the concurrency around creation and deletion of volumes due to the above issues. This added the following time to the test case:
```
creation non-parallel = ~8 additional seconds
deletion non-parallel = ~15 additional seconds
```
Otherwise, this should be good.

It was weird that `ListVolumes` was in `projects.go`, but to be consistent I added the test into `projects_test.go` file. That function/test case could probably just get moved into `volumes.go`/`volumes_test.go` and renamed `List`, but to avoid adding too much into this PR I held off.

t0mk commented:
@ericrpowers Wow, you are right, the SSH keys don't support pagination. I just learned that. thanks for listing the routes with pagination. Good idea to pivot to the volumes.

I reported the two API issues to Packet Jira, thanks.

Please do move the ListVolumes to the volumes.go under the Volumes endpoint and, call it List(projectID string, opts *ListOptions) - just like you suggested. And the test to volumes_test.go then.

I think you really don't have to care about the duration of the tests as long as it's under 5 minutes (that's an average duration of the basic device test). If I would be doing the volumes listing test, I wouldn't even use sync and waitGroups, I'd just create the test volumes sequentially because
 - as you noticed, there are API issues when creating resources in parallel
 - the duration of creating 11 volumes is close to the duration single basic device test
 - code is more simple without the synchronization

.. but it's up to you, I don't consider this important enough to argue about.

Don't worry about API changes, we will bump the version after the pagination is complete.

ericrpowers commented on volumes_test.go:
```diff
@@ -9,14 +9,14 @@ import (
 		if c.State == "active" {
 			return c, nil
 		}
+		<-time.After(5 * time.Second)
```
Did just a tiny adjustment here to save time when waiting for volumes in a loop (made the longer test case from 90s to 45s). : )

ericrpowers commented:
That should cover all your comments, but if I missed any just let me know! Also, I can revert the above adjustment if preferred ('bad' habit of mine XD).

t0mk commented:
thanks. The code looks good and the relevant tests run, I am merging.
